build ufe Python bindings module consistently with other modules

mattyjams opened the pull request:
This change is similar to #207 but applied to the ufe module. Instead of building the module as `mayaUsd/ufe.so` directly, we instead build it as `mayaUsd/ufe/_ufe.so` and bring its contents into the ufe namespace with an `__init__.py` that uses `Tf.PrepareModule()`.

Also similarly to #207, this adds a `TfScriptModuleLoader` registry function that ensures that when the ufe library is loaded as a result of importing its Python module, the core USD libraries that it depends on also get loaded.

Without the changes in this PR, the test that was added fails with an error that looks like this:
```
======================================================================
ERROR: testImportModule (testUfePythonImport.UfePythonImportTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "testUfePythonImport.py", line 35, in testImportModule
    invalidPrim = mayaUsdUfe.getPrimFromRawItem(0)
TypeError: No to_python (by-value) converter found for C++ type: pxrInternal_v0_20__pxrReserved__::UsdPrim
```

ppt-adsk commented on lib/CMakeLists.txt:
```diff
@@ -636,21 +640,25 @@ if(UFE_FOUND)
     if(IS_MACOSX OR IS_LINUX)
         mayaUsd_init_rpath(rpath "${PROJECT_NAME}")
-        mayaUsd_add_rpath(rpath "../..")
+        mayaUsd_add_rpath(rpath "../../..")
```
Is this a bug you fixed, i.e. it was simply incorrect previously?

ppt-adsk replied:
Sean made me notice the change in installation directory, which now has a ufe sub-folder, thus the need to adjust the rpath.

kxl-adsk commented on test/lib/ufe/testUfePythonImport.py:
```diff
@@ -0,0 +1,39 @@
+import unittest
+
+
+class UfePythonImportTestCase(unittest.TestCase):
```
This test is failing on windows:
```5: ======================================================================
5: ERROR: testImportModule (testUfePythonImport.UfePythonImportTestCase)
5: ----------------------------------------------------------------------
5: Traceback (most recent call last):
5:   File "testUfePythonImport.py", line 35, in testImportModule
5:     invalidPrim = mayaUsdUfe.getPrimFromRawItem(0)
5: TypeError: No to_python (by-value) converter found for C++ type: class pxrInternal_v0_20__pxrReserved__::UsdPrim
5:
5: ----------------------------------------------------------------------```

mattyjams replied:
Hmm... That's the exact issue that these changes are meant to address. It's a result of the ufe library being loaded *without* having loaded the usd library. That's what the script module loader function in moduleDeps.cpp should fix.

I still don't have access to a Windows machine on my end, so I'm not able to test this myself at the moment. Could I have you run the test with the environment variable `TF_DEBUG=TF_SCRIPT_MODULE_LOADER` set so we can verify whether or not the module is getting registered?

mattyjams replied:
I did a quick test of this on Linux setting `TF_DEBUG=TF_SCRIPT_MODULE_LOADER` in an interactive session of Maya after building this branch.

I ran this command in the script editor:
```from mayaUsd import ufe as mayaUsdUfe```

And I see this output in the terminal:
```
Registering library tf with predecessors: arch, 
Registering library trace with predecessors: arch, js, tf, 
Registering library gf with predecessors: arch, tf, 
Registering library vt with predecessors: arch, gf, tf, trace, 
Registering library work with predecessors: tf, trace, 
Registering library plug with predecessors: arch, js, tf, trace, work, 
Registering library ar with predecessors: arch, js, plug, tf, vt, 
Registering library sdf with predecessors: ar, arch, gf, tf, trace, vt, work, 
Registering library pcp with predecessors: ar, sdf, tf, trace, vt, work, 
Registering library kind with predecessors: plug, tf, 
Registering library usd with predecessors: ar, arch, kind, pcp, plug, sdf, tf, trace, vt, work, 
Registering library usdGeom with predecessors: js, plug, sdf, tf, trace, usd, vt, work, 
Registering library usdUtils with predecessors: arch, gf, sdf, tf, usd, usdGeom, 
Registering library usdSkel with predecessors: arch, gf, sdf, tf, trace, usd, usdGeom, vt, work, 
Registering library ndr with predecessors: ar, plug, sdf, tf, vt, work, 
Registering library sdr with predecessors: ar, ndr, sdf, tf, vt, 
Registering library usdShade with predecessors: ndr, sdf, sdr, tf, usd, usdGeom, vt, 
Registering library usdLux with predecessors: sdf, tf, usd, usdGeom, vt, 
Registering library usdRi with predecessors: sdf, tf, usd, usdGeom, usdLux, usdShade, vt, 
Registering library usdHydra with predecessors: tf, usd, usdShade, 
Registering library garch with predecessors: arch, tf, 
Registering library glf with predecessors: ar, arch, garch, gf, hf, js, plug, sdf, tf, trace, 
Registering library pxOsd with predecessors: gf, tf, vt, 
Registering library usdVol with predecessors: js, plug, sdf, tf, trace, usd, usdGeom, vt, work, 
Registering library cameraUtil with predecessors: gf, tf, 
Registering library usdImagingGL with predecessors: ar, garch, gf, glf, hd, hdx, hio, plug, pxOsd, sdf, sdr, tf, trace, usd, usdGeom, usdHydra, usdImaging, usdShade, vt, work, 
Registering library mayaUsd with predecessors: ar, gf, hd, hdx, js, kind, plug, sdf, tf, usd, usdGeom, usdImaging, usdImagingGL, usdLux, usdRi, usdShade, usdSkel, usdUtils, vt, 
Registering library ufe with predecessors: mayaUsd, sdf, tf, usd, usdGeom, vt, 
_LoadUpTo('ufe') {
  Load('tf');
  _LoadUpTo('tf') {
  }
  Load('trace');
  _LoadUpTo('trace') {
  }
  Load('work');
  _LoadUpTo('work') {
  }
  Load('plug');
  _LoadUpTo('plug') {
  }
  Load('gf');
  _LoadUpTo('gf') {
  }
  Load('vt');
  _LoadUpTo('vt') {
  }
  Load('ar');
  _LoadUpTo('ar') {
  }
  Load('kind');
  _LoadUpTo('kind') {
  }
  Load('sdf');
  _LoadUpTo('sdf') {
  }
  Load('pcp');
  _LoadUpTo('pcp') {
  }
  Load('usd');
  _LoadUpTo('usd') {
  }
  Load('usdGeom');
  _LoadUpTo('usdGeom') {
  }
  Load('garch');
  _LoadUpTo('garch') {
  }
  Load('glf');
  _LoadUpTo('glf') {
  }
  Load('pxOsd');
  _LoadUpTo('pxOsd') {
  }
  Load('ndr');
  _LoadUpTo('ndr') {
  }
  Load('sdr');
  _LoadUpTo('sdr') {
  }
  Load('usdShade');
  _LoadUpTo('usdShade') {
  }
  Load('usdHydra');
  _LoadUpTo('usdHydra') {
  }
  Load('usdImagingGL');
  _LoadUpTo('usdImagingGL') {
  }
  Load('usdLux');
  _LoadUpTo('usdLux') {
  }
  Load('usdRi');
  _LoadUpTo('usdRi') {
  }
  Load('usdSkel');
  _LoadUpTo('usdSkel') {
  }
  Load('usdUtils');
  _LoadUpTo('usdUtils') {
  }
  Load('mayaUsd');
  _LoadUpTo('mayaUsd') {
  }
}
```

I guess I'd hope to see something similar on Windows.

kxl-adsk replied:
```
5:  MAYA_NO_STANDALONE_ATEXIT=1
5:  TF_DEBUG=TF_SCRIPT_MODULE_LOADER
5: Test timeout computed to be: 10000000
5: MAYA_DEBUG_NO_SAVE_ON_CRASH is set.
5: Warning: At initialization event does not appear in the list of available events
5: Warning: At initialization event does not appear in the list of available events
5: Warning: At initialization event does not appear in the list of available events
5: Warning: At initialization event does not appear in the list of available events
5: Registering library tf with predecessors: arch,
5: Registering library work with predecessors: tf, trace,
5: Registering library trace with predecessors: arch, js, tf,
5: Registering library plug with predecessors: arch, js, tf, trace, work,
5: Registering library kind with predecessors: plug, tf,
5: Registering library gf with predecessors: arch, tf,
5: Registering library vt with predecessors: arch, gf, tf, trace,
5: Registering library ar with predecessors: arch, js, plug, tf, vt,
5: Registering library sdf with predecessors: ar, arch, gf, tf, trace, vt, work,
5: Registering library pcp with predecessors: ar, sdf, tf, trace, vt, work,
5: Registering library usd with predecessors: ar, arch, kind, pcp, plug, sdf, tf, trace, vt, work,
5: *** Not loading modules for unknown lib 'ufe'
5: _LoadUpTo('tf') {
5: }
5: E
5: ======================================================================
5: ERROR: testImportModule (testUfePythonImport.UfePythonImportTestCase)
5: ----------------------------------------------------------------------
5: Traceback (most recent call last):
5:   File "testUfePythonImport.py", line 35, in testImportModule
5:     invalidPrim = mayaUsdUfe.getPrimFromRawItem(0)
5: TypeError: No to_python (by-value) converter found for C++ type: class pxrInternal_v0_20__pxrReserved__::UsdPrim
5:
5: ----------------------------------------------------------------------
5: Ran 1 test in 0.184s
5:
5: FAILED (errors=1)
1/1 Test #5: testUfePythonImport ..............***Failed   12.06 sec
```

mattyjams replied:
Yeah, that seems suspicious. I would expect to see "Registering library ..." lines for both mayaUsd and ufe listed there, and there is also a big chunk of USD libraries missing as well.

Does the `testMayaUsdPythonImport` test pass for you using that build? I would expect not given what we're seeing here. In #207 where we did this for mayaUsd, I know @HamedSabri-adsk ran into an issue with the equivalent mayaUsd test whereas @fowlertADSK did not, so I wonder if something similar is happening here.

kxl-adsk replied:
`testMayaUsdPythonImport` is passing.

HamedSabri-adsk replied:
@mattyjams, I also get the exact behavior on Windows as Krystian.

mattyjams replied:
Cool, thanks for trying it @HamedSabri-adsk!

I have a hunch that it's something to do with the fact that unlike mayaUsd, there's no actual "ufe" library that gets installed (on Linux, no "lib/libufe.so"). The UFE functionality is built as just more symbols in libmayaUsd.so. I'm not sure why that is problematic only on Windows, and I'm unfortunately not able to tinker with it on that platform at the moment. :(

HamedSabri-adsk commented on test/lib/ufe/CMakeLists.txt:
```diff
@@ -103,7 +104,12 @@ foreach(script ${test_script_files})
+                                          maya.standalone.initialize(name='python'); \
+                                          import ${target}; \
+                                          main(module=${target}); \
+                                          maya.standalone.uninitialize()"
```
@kxl-adsk @mattyjams  
I recently fixed a bug in PR198 where ctest was falsely reporting python tests pass.

There are multiple issues here ( line 111,112 ):
1- main(module=${target});  by default calls the sys.exit() and therefore `maya.standalone.uninitialize()` never is called.
2- if there is an error or failure in any of these ufe tests, ctest still says that everything passed.

Solution:
1- don't let the sys.exit() happen which simply can be done by passing `exit=False` argument. 
`main(module=${target}, exit=False)`
2- call the sys.exit based on the TestProgram  program result. maya.standalone.uninitialize() should also happen before the sys.exit call.
```
maya.standalone.uninitialize();
testProg = main(module=${target}, exit=False)
sys.exit(not testProg.result.wasSuccessful());
```
Note: Interestingly, I made an error on purpose in one of these tests but to my surprise ctest reported correctly that the test fail. In any case, it still makes sense to exit based on wasSuccessful() result.

See unittest doc:
https://docs.python.org/2/library/unittest.html#unittest.main

See some of the fixes in PR198:
https://github.com/Autodesk/maya-usd/pull/198/commits/b34dc94655c33c791f360a8090a5daedc8eaee97

I am planning of creating a separate PR to fix this. I also think bundling up all these tests into a TestSuite, and then calling unittest.main once would be a better solution.

mattyjams replied:
Ahh, thanks @HamedSabri-adsk! I didn't actually change anything here other than breaking up the one long line into multiple lines. Since you're already addressing that in #198, I'll just revert this so that this PR merges cleanly with yours.

HamedSabri-adsk replied:
@mattyjams Nice, thanks for doing that.

mattyjams commented:
I rebased this on the current dev branch and reverted the change in test/lib/ufe/CMakeLists.txt to break up the long test command line, since @HamedSabri-adsk addresses that in #198 and also fixes some issues with it.

I have a request in to our systems team to see if they can get me access to a Windows box, but until then I'm unfortunately not able to dig into what's going wrong on Windows.

mattyjams commented:
Rebased this again, and the default constructed USD object issue was fixed in a yet-to-be-pushed commit in core USD, so I adapted the test to work with that fix.

Still no update on the Windows test failure though, sorry. :(

kxl-adsk commented:
Does this mean we can't merge this before updating to yet-to-be-pushed USD?

kxl-adsk commented:
As well, how about older USD versions? We said 19.07 as a minimum.

mattyjams commented:
No, with the change I made to the test, it should work with all versions of USD. Before the core commit goes in, the expected type name is "Object" and after, it's "Prim". The test now just makes sure it's one of those two:
https://github.com/mattyjams/maya-usd/commit/bf777d9fbfd85a0657f4fbafcb6f1f50b2236084

kxl-adsk commented:
Perfect, I read the comment but haven't looked at the code (my mistake). FWIW. I'm trying to find time to look at this windows failure but had no luck.

kxl-adsk commented:
@mattyjams I tried to debug this a bit. No solution unfortunately, but noticed that script module loader does the right thing when I have both lib and ufe imported, i.e. 
```
        from mayaUsd import lib as mayaUsdLib
        from mayaUsd import ufe as mayaUsdUfe
```

Then I'm getting following registration and load trace.
```
6: Registering library tf with predecessors: arch,
6: Registering library gf with predecessors: arch, tf,
6: Registering library vt with predecessors: arch, gf, tf, trace,
6: Registering library work with predecessors: tf, trace,
6: Registering library trace with predecessors: arch, js, tf,
6: Registering library plug with predecessors: arch, js, tf, trace, work,
6: Registering library garch with predecessors: arch, tf,
6: Registering library ar with predecessors: arch, js, plug, tf, vt,
6: Registering library sdf with predecessors: ar, arch, gf, tf, trace, vt, work,
6: Registering library glf with predecessors: ar, arch, garch, gf, hf, js, plug, sdf, tf, trace,
6: Registering library cameraUtil with predecessors: gf, tf,
6: Registering library pxOsd with predecessors: gf, tf, vt,
6: Registering library ndr with predecessors: ar, plug, sdf, tf, vt, work,
6: Registering library sdr with predecessors: ar, ndr, sdf, tf, vt,
6: Registering library kind with predecessors: plug, tf,
6: Registering library pcp with predecessors: ar, sdf, tf, trace, vt, work,
6: Registering library usd with predecessors: ar, arch, kind, pcp, plug, sdf, tf, trace, vt, work,
6: Registering library usdGeom with predecessors: js, plug, sdf, tf, trace, usd, vt, work,
6: Registering library usdShade with predecessors: ndr, sdf, sdr, tf, usd, usdGeom, vt,
6: Registering library usdHydra with predecessors: tf, usd, usdShade,
6: Registering library usdLux with predecessors: sdf, tf, usd, usdGeom, vt,
6: Registering library usdVol with predecessors: js, plug, sdf, tf, trace, usd, usdGeom, vt, work,
6: Registering library usdImagingGL with predecessors: ar, garch, gf, glf, hd, hdx, hio, plug, pxOsd, sdf, sdr, tf, trace, usd, usdGeom, usdHydra, usdImaging, usdShade, vt, work,
6: Registering library usdRi with predecessors: sdf, tf, usd, usdGeom, usdLux, usdShade, vt,
6: Registering library usdSkel with predecessors: arch, gf, sdf, tf, trace, usd, usdGeom, vt, work,
6: Registering library usdUtils with predecessors: arch, gf, sdf, tf, usd, usdGeom,
6: Registering library mayaUsd with predecessors: ar, gf, hd, hdx, js, kind, plug, sdf, tf, usd, usdGeom, usdImaging, usdImagingGL, usdLux, usdRi, usdShade, usdSkel, usdUtils, vt,
6: Registering library ufe with predecessors: mayaUsd, sdf, tf, usd, usdGeom, vt,
6: _LoadUpTo('mayaUsd') {
6:   Load('tf');
6:   _LoadUpTo('tf') {
6:   }
6:   Load('trace');
6:   _LoadUpTo('trace') {
6:   }
6:   Load('work');
6:   _LoadUpTo('work') {
6:   }
6:   Load('plug');
6:   _LoadUpTo('plug') {
6:   }
6:   Load('gf');
6:   _LoadUpTo('gf') {
6:   }
6:   Load('vt');
6:   _LoadUpTo('vt') {
6:   }
6:   Load('ar');
6:   _LoadUpTo('ar') {
6:   }
6:   Load('kind');
6:   _LoadUpTo('kind') {
6:   }
6:   Load('sdf');
6:   _LoadUpTo('sdf') {
6:   }
6:   Load('pcp');
6:   _LoadUpTo('pcp') {
6:   }
6:   Load('usd');
6:   _LoadUpTo('usd') {
6:   }
6:   Load('usdGeom');
6:   _LoadUpTo('usdGeom') {
6:   }
6:   Load('garch');
6:   _LoadUpTo('garch') {
6:   }
6:   Load('glf');
6:   _LoadUpTo('glf') {
6:   }
6:   Load('pxOsd');
6:   _LoadUpTo('pxOsd') {
6:   }
6:   Load('ndr');
6:   _LoadUpTo('ndr') {
6:   }
6:   Load('sdr');
6:   _LoadUpTo('sdr') {
6:   }
6:   Load('usdShade');
6:   _LoadUpTo('usdShade') {
6:   }
6:   Load('usdHydra');
6:   _LoadUpTo('usdHydra') {
6:   }
6:   Load('usdImagingGL');
6:   _LoadUpTo('usdImagingGL') {
6:   }
6:   Load('usdLux');
6:   _LoadUpTo('usdLux') {
6:   }
6:   Load('usdRi');
6:   _LoadUpTo('usdRi') {
6:   }
6:   Load('usdSkel');
6:   _LoadUpTo('usdSkel') {
6:   }
6:   Load('usdUtils');
6:   _LoadUpTo('usdUtils') {
6:   }
6: }
```

Rather surprising, since moduleDep looks to be correctly setup. I tried to rename mayaUsd to something else to avoid potential collisions but that didn't make any difference.

kxl-adsk commented:
@mattyjams How do we move this forward? I don't think we can help much more to debug it than what was provided https://github.com/Autodesk/maya-usd/pull/209#issuecomment-583108500

mattyjams commented:
Thanks for following up, @kxl-adsk. I was finally able to get my hands on a Windows box, so I'm hoping to find some time next week to fumble my way through the Windows build and see if I can get to the bottom of this.

kxl-adsk commented:
@mattyjams Were you able to reproduce the issue? Any update?

mattyjams commented:
Sorry @kxl-adsk, no update on this yet. I've been tied up with other things and didn't get a chance to look at this on Windows as I'd hoped.

What I pushed yesterday was just a rebase after #323 was merged.

kxl-adsk commented:
@mattyjams So appears like if we wait long enough things get magically fixed;) I suspect refactoring from @elrond79 which might have fixed it. Still need to make few more tests to be sure...

mattyjams commented:
> @mattyjams So appears like if we wait long enough things get magically fixed;) I suspect refactoring from @elrond79 which might have fixed it. Still need to make few more tests to be sure...

@kxl-adsk Ha! Excellent! Of course, secretly that was my plan the whole time. ;)

Otherwise, still no update from me on the Windows front. I was planning to rebase this after all of the refactoring changes go in and then give it another go. If this is looking good for you on Windows now though, definitely feel free to merge it in. We've been including this in what we build internally for a few months, but that's admittedly only on Linux, whereas Windows was the sticking point here.

kxl-adsk commented:
Once #388 is in, I will retest it again.

kxl-adsk commented:
@mattyjams please resolve conflicts and I will give it a final round of testing (and hopefully merge after)

mattyjams commented:
@kxl-adsk: Rebased! I verified that the `testUfePythonImport` test that this adds passes for me on both Linux and Mac. Let me know how it goes for you on Windows!
